Add wasm32-unknown-emscripten platform support document (#131582)
rust-lang/rust: src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md

integrates with JavaScipt on the web to provide a more featureful standard library.
This target is often used when compiling for the web, accessing web APIs, or
interoperating with C and C++ code.

One existing user of this target is the
[`pyodide` project](https://pyodide.org/) which provides a Python runtime in
WebAssembly using Emscripten and compiles Python extension modules written in Rust
to the `wasm32-unknown-emscripten` target.

If you want to target the web but no not need interop with C/C++ or you want to
produce a bare-bones WebAssembly binary and you do not require access to the host
environment (e.g. stdio, file access, networking, etc.), consider using the
[`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target instead.

If you need access to the host environment, consider using one of the WASI targets
([`wasm32-wasip1`](./wasm32-wasip1.md) or [`wasm32-wasip2`](./wasm32-wasip2.md)).

## Target maintainers

When this target was added to the compiler, platform-specific documentation here
was not maintained at that time. This means that the list below is not
exhaustive, and there are more interested parties in this target. That being

> said, those interested in maintaining this target are:

Reviewer: I realize this is probably copy/pasted from the `wasm32-unknown-unknown` target but given the set of three maintainers below you can probably drop this clause. I mostly added it to the unknown target because I was the only one listed and wanted to clarify why there was just one.

Author: Done

$ file foo.wasm
```

## Cross-compilation

This target can be cross-compiled from any host.

## Emscripten ABI Compatibility

The Emscripten compiler toolchain does not follow a semantic versioning scheme
that clearly indicates when breaking changes to the ABI can be made. If you are
using the pre-compiled `std` from rustup, it may be compiled with a different
Emscripten version than the local Emscripten you have installed. If a breaking
change to the ABI was made in between the two versions, your code will exhibit
undefined behaviour. If you come across issues, you can rebuild the Rust standard
library with your local Emscripten version using

```sh
$ cargo +nightly -Zbuild-std build
```

## Testing

This target is not extensively tested in CI for the rust-lang/rust repository.
It's recommended to test the [`wasm32-wasip1`](./wasm32-wasip1.md) target instead
for WebAssembly compatibility.

## Conditionally compiling code

It's recommended to conditionally compile code for this target with:

```text
#[cfg(all(target_family = "wasm", target_os = "emscripten"))]
```

Note that there is no way to tell via `#[cfg]` whether code will be running on

> the web or not.

Reviewer: This clause may not be so applicable for the emscripten target since by definition almost all emscripten code runs on the web

Author: Removed - done

## Enabled WebAssembly features

WebAssembly is an evolving standard which adds new features such as new
instructions over time. The `wasm32-unknown-emscripten` target inherits the

> default settings of Emscripten.

Reviewer: This might want to be clarified a bit because it's not quite true. The default features are inherited from LLVM that rustc uses which may differ from Emscripten depending on the user's installation of the Emscripten toolchain. In that sense I'd recommend pointing to the `wasm32-unknown-unknown` docs about features here because the documentation there will apply to the Emscripten target here too.

Author: I added a reference to the `wasm32-unknown-unknown` docs

support interop with C/C++ (and other languages), and support most of the Rust
standard library. While the WASI targets are portable across different hosts
(web and non-web), WASI has no standard way of accessing web APIs, whereas
Emscripten has the ability to run arbitrary JS from WASM and access many web APIs.
If you are only targeting the web and need to access web APIs, the
`wasm32-unknown-emscripten` target may be preferable.

## Target maintainers

- Hood Chatham, https://github.com/hoodmane
- Sam Clegg, https://github.com/sbc100
- Juniper Tyree, https://github.com/juntyr

## Requirements

This target is cross-compiled. The Emscripten compiler toolchain (emsdk) must be
installed to link Rust-compiled code with Emscripten's libc and other C/C++
code. Please refer to <https://emscripten.org/docs/getting_started/downloads.html>
on how to get started with the emsdk.

## Building the target

Building this target can be done by:

* Configure the `wasm32-unknown-emscripten` target to get built.
* Configure LLD to be built.
* Ensure the `WebAssembly` target backend is not disabled in LLVM.

These are all controlled through `config.toml` options. It should be possible

> to build this target on any platform.

Reviewer 2: Suggestion: could you show a minimal `config.toml` snippet for this?

Author: I've tried my best to add a minimal example config.toml

## Building Rust programs

Rust programs can be compiled by adding this target via rustup:

```sh
$ rustup target add wasm32-unknown-emscripten
```

and then compiling with the target:

```sh
$ rustc foo.rs --target wasm32-unknown-emscripten
$ file foo.wasm
```

## Cross-compilation

This target can be cross-compiled from any host.

## Emscripten ABI Compatibility

The Emscripten compiler toolchain does not follow a semantic versioning scheme
that clearly indicates when breaking changes to the ABI can be made. If you are
using the pre-compiled `std` from rustup, it may be compiled with a different
Emscripten version than the local Emscripten you have installed. If a breaking
change to the ABI was made in between the two versions, your code will exhibit
undefined behaviour. If you come across issues, you can rebuild the Rust standard
library with your local Emscripten version using

```sh
$ cargo +nightly -Zbuild-std build
```

## Testing

This target is not extensively tested in CI for the rust-lang/rust repository.
It's recommended to test the [`wasm32-wasip1`](./wasm32-wasip1.md) target instead

> for WebAssembly compatibility.

Reviewer 2: Could you elaborate how contributors can test this target locally?

i.e. Tier 2 target policy says

> The target must provide documentation for the Rust community explaining [..] how to run tests for the target. If at all possible, this documentation should show how to run Rust programs and tests for the target using emulation, to allow anyone to do so. If the target cannot be feasibly emulated, the documentation should explain how to obtain and work with physical hardware, cloud systems, or equivalent.

see #131666, where Jubilee found that

> a fifth of the rustc UI test suite to fail on wasm32-unknown-emscripten when trying to run ./x.py test --target wasm32-unknown-emscripten

Reviewer 2: I was able to get past the test helpers stage with <https://emscripten.org/docs/getting_started/downloads.html> including to clone that repo, use `./emsdk` to fetch and activate the latest emcc toolchain, then use a provided shell script to setup the relevant env vars and make them available to bootstrap.

stage1 std seems to be wonky in the bootstrap test setup

```
warning: dropping unsupported crate type `dylib` for target `wasm32-unknown-emscripten`

warning: `std` (lib) generated 1 warning
```

ui tests failed due to linker-flavors and such and I didn't bother investigating.

Author: @hoodmane Since you have gotten most tests to run, could you suggest some text?

Reviewer 3: I've been running tests with:
```sh
./x.py test --target wasm32-unknown-emscripten --skip src/tools/linkchecker
```

Reviewer 3: But you need `emcc` and `node` on the path first. To get `emcc` installed and on the path:
```sh
git clone https://github.com/emscripten-core/emsdk.git --depth 1
./emsdk/emsdk install 3.1.68
./emsdk/emsdk activate 3.1.68
source ./emsdk/emsdk_env.sh
```
And you can use whatever normal way to install node, but if we're going to recommend something, nvm is quite nice:
https://github.com/nvm-sh/nvm#install--update-script

Author: Thanks! I've added those instructions

## Conditionally compiling code

It's recommended to conditionally compile code for this target with:

```text

> #[cfg(all(target_family = "wasm", target_os = "emscripten"))]

Reviewer 3: I'd say it's sufficient to do:
```
#[cfg(target_os = "emscripten")]
```
It's often necessary to filter for wasm, not emscripten somewhere else:
```
#[cfg(all(target_family = "wasm", not(target_os = "emscripten)))]` 
```
I've hit a bunch of duplicate definition compile errors, though if people are building for Emscripten at all they would catch this one.

In a lot of cases the Emscripten branch is the same as the linux branch since Emscripten implements a lot of linux headers.

Author: I've updated the text - thanks!

```

## Enabled WebAssembly features

WebAssembly is an evolving standard which adds new features such as new
instructions over time. This target's default set of supported WebAssembly
features will additionally change over time. The `wasm32-unknown-emscripten` target
inherits the default settings of LLVM which typically, but not necessarily, matches

> the default settings of Emscripten as well.

Reviewer 3: > inherits the default settings of LLVM

I would not say this, at least not at link time. `emcc` invokes the linker with a very large number of special options.

Author: I've adapted the text

Please refer to the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md)
target's documentation on which WebAssembly features Rust enables by default, how
features can be disabled, and how Rust code can be conditionally compiled based on

> which features are enabled.

Reviewer 3: Over there, they list:
* `multivalue`
* `mutable-globals`
* `reference-types`
* `sign-ext`

These are all determined at compile time, so the advice over there should work. It does not apply as well to js/wasm interop features which are generally determined at link-time.

We might also note here that there is no support for `-fwasm-exceptions`, only for `-fexceptions` which will be added by default unless you compile with `panic=abort`. This is an important limitation. I would really like support for `wasm-exceptions`. 
https://github.com/rust-lang/rust/issues/112195

It would be nice to add wasm/js bigint interop (`-sWASM_BIGINT`) as another default feature. It's been around much longer than reference types and it makes a big improvement in code generation.

Author: I agree with enabling the `-sWASM_BIGINT` feature. This should be tracked in a separate PR.

Author: I've added a paragraph on the exception handling features.

Reviewer 3: Opened a PR for `-sWASM_BIGINT`
https://github.com/rust-lang/rust/pull/131736

# `wasm32-unknown-emscipten`

**Tier: 2**

The `wasm32-unknown-emscripten` target is a WebAssembly compilation target which
uses the [Emscripten](https://emscripten.org/) compiler toolchain. Emscripten is
a C/C++ toolchain used for running C/C++ code on the web. It provides POSIX `libc`

> and `libstd` implementations, access to the OpenGL and SDL APIs, and the ability

Reviewer 3: It implements most of posix in a mostly compatible way, and also a lot of linux APIs. It might be worth saying here that the design goal is to allow C/C++ code (e.g., 3d games) written to target linux to be ported with as few changes as possible. Also, it uses musl libc.

Author: I've adapted the wording.

# `wasm32-unknown-emscipten`

**Tier: 2**

The `wasm32-unknown-emscripten` target is a WebAssembly compilation target which
uses the [Emscripten](https://emscripten.org/) compiler toolchain. Emscripten is
a C/C++ toolchain designed to make it as easy as possible to port C/C++ code
written for Linux to run on the web. It thus provides POSIX-compatible (musl) `libc`
and `libstd` implementations and many Linux APIs, access to the OpenGL and SDL APIs,
and the ability to run arbitrary JavaScript code, all based on web APIs using JS
glue code. With the `wasm32-unknown-emscripten` target, Rust code can interoperate

> with Emscripten's ecosystem, C/C++ and JS code, and web APIs.

Reviewer 3: It would be nice if we could mention somewhere in here that emscripten code also can run in node or other JavaScript runtimes. (The goal of Emscripten is to run on web, but once you run on web it's not too hard to also run in other JS runtimes.)

Author: Done

One existing user of this target is the
[`pyodide` project](https://pyodide.org/) which provides a Python runtime in
WebAssembly using Emscripten and compiles Python extension modules written in Rust
to the `wasm32-unknown-emscripten` target.

> If you want to generate a standalone bare-bones WebAssembly binary that does not
> require JS glue code and access to the web APIs and can thus run outside the web,
> the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target may be better
> suited for you. However, [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md)
> only has partial support for the Rust standard library and does not (easily)
> support interop with C/C++ code.

Reviewer 3: ```suggestion
If you want to generate a standalone WebAssembly binary that does not
require access to the web APIs or the Rust standard library,
the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target may be better
suited for you. However, [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) does not (easily)
support interop with C/C++ code.
```
As an aside, someday someone should really add an option to make `wasm32-unknown-unknown` emit the clang ABI instead of the wasm-bindgen ABI to fix the interop problem...

Like Emscripten, the WASI targets [`wasm32-wasip1`](./wasm32-wasip1.md) and
[`wasm32-wasip2`](./wasm32-wasip2.md) also provide access to the host environment,
support interop with C/C++ (and other languages), and support most of the Rust
standard library. While the WASI targets are portable across different hosts
(web and non-web), WASI has no standard way of accessing web APIs, whereas
Emscripten has the ability to run arbitrary JS from WASM and access many web APIs.
If you are only targeting the web and need to access web APIs, the
`wasm32-unknown-emscripten` target may be preferable.

## Target maintainers

- Hood Chatham, https://github.com/hoodmane
- Sam Clegg, https://github.com/sbc100
- Juniper Tyree, https://github.com/juntyr

## Requirements

> This target is cross-compiled. The Emscripten compiler toolchain (emsdk) must be

Reviewer 3: The emscripten compiler toolchain is `emcc`, `emsdk` is the "Emscripten software developer kit" which is primarily an installer and version manager for the emscripten compiler toolchain.

Author: Thanks for catching that

installed to link WASM binaries with `emcc`. Please refer to
<https://emscripten.org/docs/getting_started/downloads.html> on how to get started
with the emsdk.

> You can also build the emsdk from source with:

Reviewer 3: This does not build from source. emsdk is just an installer for the toolchain.

Interestingly, [the code for the emscripten compiler toolchain](https://github.com/emscripten-core/emscripten/) itself has no compilation step, it is implemented in a mixture of Python and JavaScript. So you can just modify the source files in place and immediately have a patched compiler. The standard library / sysroot is compiled as needed (since `emcc` implements many many different ABIs it doesn't make sense to ship prebuilt static libraries, it builds them as needed and caches them).

All of the compiled code needed for implementing the compiler (as opposed to host wasm code) is upstream either in [binaryen](https://github.com/WebAssembly/binaryen) or in llvm itself.

Author: Thanks for correcting my mistake - I've reworded the paragraph

These are all controlled through `config.toml` options. It should be possible
to build this target on any platform. A minimal example configuration would be:

```toml
[llvm]
targets = "WebAssembly"

[build]
build-stage = 1
target = ["wasm32-unknown-emscripten"]
```

## Building Rust programs

Rust programs can be compiled by adding this target via rustup:

```sh
$ rustup target add wasm32-unknown-emscripten
```

and then compiling with the target:

```sh
$ rustc foo.rs --target wasm32-unknown-emscripten
$ file foo.wasm
```

## Cross-compilation

This target can be cross-compiled from any host.

## Emscripten ABI Compatibility

The Emscripten compiler toolchain does not follow a semantic versioning scheme
that clearly indicates when breaking changes to the ABI can be made. If you are
using the pre-compiled `std` from rustup, it may be compiled with a different
Emscripten version than the local Emscripten you have installed. If a breaking
change to the ABI was made in between the two versions, your code will exhibit
undefined behaviour. If you come across issues, you can rebuild the Rust standard
library with your local Emscripten version using

> 

Reviewer 3: The situation is actually worse because Emscripten offers many different ABIs even for a single version of Emscripten depending on linker flags. To be ABI compatible, all of the linker flags that adjust ABIs must match. This includes things like `-fexceptions` and `-sWASM_BIGINT`. I think it would be good to be a bit more emphatic here. Best practices is either:
1. use `-Zbuild-std`, or
2. match the emscripten version bundled with Rust and be careful about any `-C link-arg`'s that you pass.

The third yolo option is not ideal...

Author: I made some further adjustments to the text to highlight that almost everybody should use `-Zbuild-std`

to build this target on any platform. A minimal example configuration would be:

```toml
[llvm]
targets = "WebAssembly"

[build]
build-stage = 1
target = ["wasm32-unknown-emscripten"]
```

## Building Rust programs

Rust programs can be compiled by adding this target via rustup:

```sh
$ rustup target add wasm32-unknown-emscripten
```

and then compiling with the target:

```sh
$ rustc foo.rs --target wasm32-unknown-emscripten
$ file foo.wasm
```

## Cross-compilation

This target can be cross-compiled from any host.

## Emscripten ABI Compatibility

The Emscripten compiler toolchain does not follow a semantic versioning scheme
that clearly indicates when breaking changes to the ABI can be made. Additionally,
Emscripten offers many different ABIs even for a single version of Emscripten
depending on the linker flags used, e.g. `-fexceptions` and `-sWASM_BIGINT`. If
the ABIs mismatch, your code may exhibit undefined behaviour.

To ensure that the ABIs of your Rust code, of the Rust standard library, and of
other code compiled for Emscripten all match, you should rebuild the Rust standard
library with your local Emscripten version and settings using:

```sh

> cargo +nightly -Zbuild-std build

Reviewer 3: We should perhaps add a note here that this won't work for crates that use:
1. cargo vendor
2. panic=abort

You can work around the cargo vendor problem by building a custom sysroot with https://github.com/ralfJung/rustc-build-sysroot/ you still won't be able to build crates with `panic=abort` .

# `wasm32-unknown-emscipten`

**Tier: 2**

The `wasm32-unknown-emscripten` target is a WebAssembly compilation target which
uses the Emscripten compiler toolchain. Unlike the
[`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target, which only has
partial support for the Rust standard library, `wasm32-unknown-emscripten`
integrates with JavaScipt on the web to provide a more featureful standard library.
This target is often used when compiling for the web, accessing web APIs, or

> interoperating with C and C++ code.

Reviewer 4: Should we mention something about preferring `wasm32-unknown-unknown` or the `wasi` targets if C/C++ interop isn't needed? I'm not sure what kind of usecases would encourage using one over the other, but this would be great to document to steer people to the right thing.

Cc @alexcrichton

Author: I've added two paragraphs at the end of the section but would also appreciate Alex's input.

Reviewer: I might recommend rephrasing/reorganizing this section along the lines of:

1. What is Emscripten and how this target integrates? (e.g. Emscripten is a C/C++ toolchain not related to Rust used for running C/C++ code on the web)
2. Why Emscripten is used and how Rust integrates. (e.g. Emscripten implements POSIX in terms of web APIs and provides a C standard library for that)
3. How this contrasts to other wasm targets in Rust (e.g. unlike `wasm32-unknown-unknown` it can run C/C++ and provides a full implementation of libc/libstd, unlike `wasm32-wasip1` it requires JS glue code to run on the web because the interface between wasm/JS isn't standard)

In terms of C/C++ I'd recommend focusing on both that and library support for Emscripten. Emscripten isn't just about C/C++ but also all the polyfills provided for libraries like SDL (IIRC). The contrast to `wasm32-unknown-unknown` can indicate that the unknown target has no support for C/C++ or anything that requires JS (or at least not inherently).

When comparing to `wasm32-wasip1` I'd recommend mostly contrasting it in terms of web support. The WASI target has no standard way of accessing web APIs where Emscripten has the ability to run arbitrary JS from wasm and access many web APIs. In the end the difference here will likely boil down to "are you running on the web or not" but that doesn't alone accurately capture the nuance of you can run WASI on the web but you just won't have access to the full suite of web APIs by default.

---

That may be more words than desired, sorry about that! Overall I think the wording is fine as-is, so what I'm thinking here is just minor updates here and there. Feel free to take or leave any suggestions as you see fit!

Author: Thank you for your extensive feedback @alexcrichton, I really appreciate it!

I've tried to address it in the latest commit, I hope I managed to capture your feedback

One existing user of this target (please feel free to edit and expand this list

> too) is the [`pyodide` project](https://github.com/pyodide/pyodide) which provides

Reviewer 4: Not sure if the `please feel free to edit and expand this list too` part was intentional but it can probably be removed.

Author: Done

a Python runtime in WebAssembly using Emscripten and compiles Python extension
modules written in Rust to the `wasm32-unknown-emscripten` target.

## Target maintainers

When this target was added to the compiler, platform-specific documentation here
was not maintained at that time. This means that the list below is not
exhaustive, and there are more interested parties in this target. That being
said, those interested in maintaining this target are:

- TODO

## Requirements

This target is cross-compiled. The Emscripten compiler toolchain must be
installed to link Rust-compiled code with Emscripten's libc and other C/C++

> code.

Reviewer 4: Something here should probably link to https://emscripten.org/, https://emscripten.org/docs/getting_started/downloads.html, or the relevant github repos.

Author: Done
